File: GFG_streak/KMP-algo.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
          // onion
        // 00012
        // m m m m m m
        // m m
        // o n i o n i o l o n i o n
        // 0 0 0 1 2 3 4 0 1 2 i
        // m m m
        // m m m
        // a a b a a a c
        // 0 1 0 1 2 0 0
        // m m
        // a a b a a a c
        // 0 1 i
        //  m
        //  e d u
        //  0 0 0
    vector<int> ComputeLPS(string &pat){
        int n = pat.size();
        vector<int> lps (n, 0);
        int i=1, match=0;
        while(i < n){
            // cout << pat[i] << match << pat[match] << endl;
            if(pat[i] == pat[match]){
                lps[i++] = ++match;
            }
            else{
                // if(match == 0){
                //     // lps[i++] = 0;
                //     i++;
                // }
                // else{
                //     match = match-1;
                // }

                if(match!=0){
                    match = lps[match-1];
                }
                else{
                    i++;
                }
            }
        }

        return lps;
    }
    vector<int> search(string& pat, string& txt) {
        // code here

        vector<int> lps = ComputeLPS(pat);
        vector<int> ans;
        int n = txt.size(), pn = pat.size(), i = 0, j=0;
        while(i < n){
            if(txt[i] == pat[j]){
                i++;
                j++;
            }
            else{
                if(j!=0){
                    j = lps[j-1];
                }
                else{
                    i+=1;
                }
            }

            if(j==pn){
                ans.push_back(i-j);
                j = lps[j-1];
            }
            // i++;
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string S, pat;
        cin >> S >> pat;
        Solution ob;
        vector<int> res = ob.search(pat, S);
        if (res.size() == 0)
            cout << "[]" << endl;
        else {
            for (int i : res)
                cout << i << " ";
            cout << endl;
        }
    }
    return 0;
}

// } Driver Code Ends

File: LC_PYTHON/GFG_streak/KMP-algo.py
#User function Template for python3

class Solution:
    def ComputeLPS(self, pat):
        lps = [0]*len(pat)
        i = 1
        match = 0
        while i < len(pat):
            if pat[i] == pat[match] :
                lps[i] = match+1
                i+=1
                match+=1
            else:
                if match !=0:
                    match-=1
                else:
                    i+=1
        return lps
                    
    def search(self, pat, txt):
        # code here
        lps = self.ComputeLPS(pat)
        ans = []
        i = 0
        j = 0
        while i < len(txt):
            if txt[i] == pat[j]:
                i+=1
                j+=1
            else:
                if j!=0:
                    j = lps[j-1]
                else:
                    i+=1
            if j==len(pat):
                ans.append(i-j)
                j = lps[j-1]
        return ans

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        s = input().strip()
        patt = input().strip()
        ob = Solution()
        ans = ob.search(patt, s)
        if len(ans) == 0:
            print("[]", end="")
        for value in ans:
            print(value, end=' ')
        print()

# } Driver Code Ends